tidb-binlog/driver/reader/kafka_consumer/kafka_go.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafka_consumer
+
+import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/segmentio/kafka-go"
+	"go.uber.org/zap"
+)
+
+const (
+	KafkaGOType = "kafka-go"
+)
+
+type KafkaGO struct {
+	// client is high level api, only can consumer from offset
+	client *kafka.Reader
+	// conn is low level api, which has createTopic DeleteTopic and other more function than *kafka.Reader
+	conn *kafka.Conn
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// TODO reconnect *kafka.Conn if broker is down
+func newConn(ctx context.Context, cfg *KafkaConfig) (conn *kafka.Conn, err error) {
+	for _, addr := range cfg.Addr {
+		conn, err = kafka.DialLeader(ctx, "tcp", addr, cfg.Topic, int(cfg.Partition))
+		if err == nil {
+			return
+		}
+	}
+	return
+}
+
+// NewKafkaGoConsumer return kafka-go consumer on specify topic and partition
+func NewKafkaGoConsumer(cfg *KafkaConfig) (Consumer, error) {
+	if len(cfg.Addr) == 0 {
+		return nil, errors.New("no available kafka address")
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	conn, err := newConn(ctx, cfg)
+	if err != nil {
+		cancel()
+		return nil, errors.Trace(err)
+	}
+	return &KafkaGO{
+		client: kafka.NewReader(kafka.ReaderConfig{
+			Brokers:   cfg.Addr,
+			Topic:     cfg.Topic,
+			Partition: int(cfg.Partition),
+			// MinBytes and MaxBytes define the size of response in fetch request
+			// if one message is large than MaxBytes, it will consume in several batches
+			MinBytes: 10e3, // 1KB
+			MaxBytes: 10e6, // 1MB
+		}),
+		conn:   conn,
+		ctx:    ctx,
+		cancel: cancel,
+	}, nil
+}
+
+// ConsumerFromOffset implements Consumer.ConsumerFromOffset
+func (k *KafkaGO) ConsumeFromOffset(offset int64, consumerChan chan<- *KafkaMsg, done <-chan struct{}) error {
```

Comment:
consider to discard `done <-chan struct{}` argument

Reply:
moved